python_modules/dagster/dagster/_core/definitions/source_asset.py

```diff
@@ -187,6 +178,9 @@ def _get_op_def_compute_fn(self, observe_fn: SourceAssetObserveFunction):
             DecoratedOpFunction,
             is_context_provided,
         )
+        from dagster._core.execution.context.compute import (
```

Comment:
is this file an unrelated change?

Reply:
No. The old code fails with the new method for subtle reasons. When `OpExecutionContext` is defined in a `TYPE_CHECKING` block, it does not exist at runtime. When accessing this annotation through `inspect.signature`, we get a string `"OpExecutionContext"` with no attempt to evaluate because `OpExecutionContext` does not correspond to an input/output/config arg, so it does not fail.

`typing.get_type_hints` attempts to evaluate `OpExecutionContext`, and raises a `NameError`. We avoid this by making it an actual type.

This _is_ technically breaking for users who have done something similar (annotated a context arg with a string referring to a type that is only imported in a `TYPE_CHECKING` block), although it seems unlikely people are doing this since string annotations with ops/asset params will already cause breakage most of the time.

That said I could add some code to catch this error and fall back to old behavior.

Reply:
Even if its unlikely its seems like a very terrible error experience so seems worth avoiding. Lets get this case explicitly under test to demonstrate how it errors. 

Should we just try to do the string lookup in globals for the type when we end up with a string from `inspect` instead of eagerly running `get_type_hints`?

Reply:
Its hard to concoct scenarios where an end user would hit the `OpExecutionContext` problem specifically, but it does seem worth avoiding introducing a known breakage

Reply:
>Should we just try to do the string lookup in globals for the type when we end up with a string from inspect instead of eagerly running get_type_hints?

I think `globals()` would be insufficient because it returns the dict for the _current_ module namespace. We need access to the dict for the module namespace where the signature we're processing is written-- `get_type_hints` must access this under the hood somehow.

I think `try`-`except` block that catches `NameError` and falls back to the old behavior if `get_type_hints` fails is the way to go. Note that this will still make code like this (which crashes now) crash:

```
from __future__ import annotations

from dagster import op
from some_lib import SomeType

from typing import TYPE_CHECKING:
     from dagster import OpExecutionContext

@op
def foo(context: OpExecutionContext, bar: SomeType):
    ...
```

This will fail because, when we fall back to the old behavior, `SomeType` will no longer be dereferenced and it will be fed to Dagster machinery as a string.

Reply:
got it - if its possible to throw a better error that would guide users in a helpful direction that would be great

Reply:
OK, I pushed a commit that catches `NameError` in `get_type_hints` and gives a thorough error message. Given this error message and the likely extreme rarity of the situation that fallback logic would fix, I don't think it's worth adding that logic.

To spell it out, for the fallback logic to save the user:

1. they would have to  have _only_ the context argument be a string and no other input/output config (which is impossible using `from __future__ import annotations`)
2. have the context type import in a `TYPE_CHECKING` block or local scope

This is very unlikely because the user is almost always going to have other runtime imports from dagster like `op`/`asset`.